mixer/adapter/rbac/rbacStore.go

```diff
@@ -0,0 +1,233 @@
+	}
+
+	serviceName := inst.Action.Service
+	if serviceName == "" {
+		env.Logger().Errorf("Missing service")
+		return false, nil
+	}
+
+	path := inst.Action.Path
+	if path == "" {
+		env.Logger().Errorf("Missing path")
+		return false, nil
+	}
+
+	method := inst.Action.Method
+	if method == "" {
+		env.Logger().Errorf("Missing method")
+		return false, nil
+	}
+
+	extraAct := inst.Action.Properties
+
+	user := inst.Subject.User
+	groups := inst.Subject.Groups
+
+	instSub := inst.Subject.Properties
+
+	rn := rs.roles[namespace]
+	if rn == nil {
+		return false, nil
+	}
+
+	for rolename, roleInfo := range rn {
+		eligibleRole := false
+		env.Logger().Infof("Checking role: %s", rolename)
+		rules := roleInfo.info.GetRules()
+		for _, rule := range rules {
+			services := rule.GetServices()
+			paths := rule.GetPaths()
+			methods := rule.GetMethods()
+			constraints := rule.GetConstraints()
+			env.Logger().Infof("Checking rule: services %v, path %v, method %v, constraints %v", services, paths, methods, constraints)
+
+			if stringMatch(serviceName, services) &&
+				(paths == nil || stringMatch(path, paths)) &&
+				stringMatch(method, methods) &&
+				checkConstraints(extraAct, constraints) {
+				eligibleRole = true
+				break
+			}
+		}
+		if !eligibleRole {
+			env.Logger().Infof("role %s is not eligible", rolename)
+			continue
+		}
+		env.Logger().Infof("role %s is eligible", rolename)
+		bindings := roleInfo.bindings
+		for _, binding := range bindings {
+			subjects := binding.GetSubjects()
+			for _, subject := range subjects {
+				if subject.GetUser() != "" && subject.GetUser() != user {
+					continue
+				}
+				if subject.GetGroup() != "" && subject.GetGroup() != groups {
+					continue
+				}
+				if checkSubject(instSub, subject.GetProperties()) {
+					return true, nil
+				}
+			}
+		}
+	}
+	return false, nil
+}
+
+// Helper function to check if a string is in a list.
+// We support four types of string matches:
+// 1. Exact match.
+// 2. Wild character match. "*" matches any string.
+// 3. Prefix match. For example, "book*" matches "bookstore", "bookshop", etc.
+// 4. Suffix match. For example, "*/review" matches "/bookstore/review", "/products/review", etc.
+func stringMatch(a string, list []string) bool {
+	for _, s := range list {
+		if a == s || s == "*" || prefixMatch(a, s) || suffixMatch(a, s) {
+			return true
+		}
+	}
+	return false
+}
+
+// Helper function to check if string "a" prefix matches "pattern".
+func prefixMatch(a string, pattern string) bool {
+	pattern = strings.TrimSuffix(pattern, "*")
+	return strings.HasPrefix(a, pattern)
+}
+
+// Helper function to check if string "a" prefix matches "pattern".
+func suffixMatch(a string, pattern string) bool {
+	pattern = strings.TrimPrefix(pattern, "*")
+	return strings.HasSuffix(a, pattern)
+}
+
+// Helper function to check if a given string value
+// is in a list of strings.
+func valueInList(value interface{}, list []string) bool {
+	switch value.(type) {
```

Comment:
do we need a switch statement here? 

```
if str, ok := value.(string); ok {
  return stringMatch(str, list)
}
return false
```

Reply:
Changed to the way you suggested. Thanks.